README.md
# Mail

- [IMAP](#imap)
- [POP3](#pop3)
- [SMTP](#smtp)

Eden sports the three common ways to interact with mail servers including SMTP, IMAP and POP3. Accessing each protocol can be acheived like in `Figure 1`.

**Figure 1. Factory**

	$imap = eden('mail')->imap(
		'imap.gmail.com', 
		'your_email@gmail.com', 
		'[YOUR PASSWORD]', 
		993, 
		true);

	$pop3 = eden('mail')->pop3(
		'pop.gmail.com', 
		'your_email@gmail.com', 
		'[YOUR PASSWORD]', 
		995, 
		true);
	
	$smtp = eden('mail')->smtp(
		'smtp.gmail.com', 
		'your_email@gmail.com', 
		'[YOUR PASSWORD]', 
		465, 
		true);

For all three protocols there are four requirements that's needed and all requirements are dependent on your specific email provider. In the figure above we use GMAIL settings as an example. The last argument is a flag. Set it to true if your email provider requires SSL.

<a name="imap"></a>
## IMAP

*Eden's* IMAP object does uses `fsocket()` rather than PHP's built in IMAP functions. We chose to use `fsocket()` instead purely because of performance reasons. You'll find setting up IMAP is overall easier with *Eden*. `Figure 2` shows how to set up the IMAP object.

**Figure 2. Setting up the IMAP Object**

	$imap = eden('mail')->imap(
		'imap.gmail.com', 
		'your_email@gmail.com', 
		'[YOUR PASSWORD]', 
		993, 
		true);

Very simply, there are four requirements that's needed and all requirements are dependent on your specific email provider. In the figure above we use GMAIL settings as an example. The last argument is a flag. Set it to true if your email provider requires SSL.

### Mailboxes

The next thing we want to know are the available mailboxes we can set to active. To get a list of mailboxes follow `Figure 3`.

**Figure 3. Mailbox List**

	$mailboxes = $imap->getMailboxes(); 

Executing the above command will give you the following results.

	Array (
		[0] => Deleted Messages
		[1] => Drafts
		[2] => INBOX
		[3] => Junk E-mail
		[4] => Notes
		[5] => Sent Messages
		[6] => Trash
		[9] => [Gmail]/All Mail
		[10] => [Gmail]/Drafts
		[11] => [Gmail]/Important
		[12] => [Gmail]/Personal
		[13] => [Gmail]/Sent Mail
		[14] => [Gmail]/Spam
		[15] => [Gmail]/Starred
		[16] => [Gmail]/System
		[17] => [Gmail]/Trash
		[18] => [Gmail]/Unsorted
	)

Every mail server we looked at has the *INBOX* folder. So when in doubt, we would say this should be the default active mailbox. Next let's set the *INBOX* as the active mailbox. `Figure 4` shows how we would set and retrieve the active mailbox in one line.

**Figure 4. Set Active Mailbox FIRST!**

	echo $imap->setMailboxes('INBOX')->getActiveInbox(); //--> INBOX 

### Emails

Now that we have set the active mailbox, we can now continue to get a list of emails. `Figure 5` shows how we would go about doing that.

**Figure 5. Getting a List of Emails**

	$emails = $imap->getEmails(0, 3); 
	$count = $imap->getEmailTotal(); 

Executing the above figure would yield you results similar to the snippet below.

**Figure 5a. Example Email Results**

	Array (
		[0] => Array
			(
				[id] => <50041ab1c9383_178f6b3294527919f@job01.tmail>
				[parent] => 
				[topic] => Trending Startups and Updates
				[mailbox] => INBOX
				[uid] => 22363
				[date] => 1342446257
				[subject] => Trending Startups and Updates
				[from] => Array( [name] => AngelList [email] => noreply@angel.co )
				[flags] => Array ( [0] => seen )
				[to] => Array ( [0] => Array( [email] => youremail@gmail.com ) )
				[cc] => Array()
				[bcc] => Array()
			)
		 
		...
	)

As you might realize, the email list does not come with the full body nor the attachments. Standard *IMAP* practices explain you should not try to get a detailed list, because the mere size of the full data could take a long while. The idea for when you should retrieve the full details is when a user specifically call to action to see the body. `Figure 6` shows how we would get the details of an email.

**Figure 6. Get the Complete Email Details**

	$email = $imap->getUniqueEmails(22363, true); 
	echo $email['body']['text/html'];

In the figure above, we call `getUniqueEmails()` the first argument should be the unique identifier (UID) found in `Figure 5a`. The second argument is whether if you want the body, leave this true usually. You could get a group of emails comma separated by UIDs in the following manner.

**Figure 6a. Get More Than One Email Detail**

	$emails = $imap->getUniqueEmails(array(22363, 22364), true);

Including the body and attachments in a list request would definitely take a long time and is not recommended.

### Searching

Searching emails follows the basic IMAP format for searching. An example of how to search can be found in `Figure 7`.

Figure 7. Searching

	$emails = $imap->search(array('TO "youremail@gmail.com"'), 0, 3); 

Search is confined to the emails in the active mailbox. This is an IMAP standard. There is no work around for this.

The combinations of search queries can be referenced below:

	ALL - return all messages matching the rest of the criteria
	ANSWERED - match messages with the \\ANSWERED flag set
	BCC "string" - match messages with "string" in the Bcc: field
	BEFORE "date" - match messages with Date: before "date"
	BODY "string" - match messages with "string" in the body of the message
	CC "string" - match messages with "string" in the Cc: field
	DELETED - match deleted messages
	FLAGGED - match messages with the \\FLAGGED (sometimes referred to as Important or Urgent) flag set
	FROM "string" - match messages with "string" in the From: field
	KEYWORD "string" - match messages with "string" as a keyword
	NEW - match new messages
	OLD - match old messages
	ON "date" - match messages with Date: matching "date"
	RECENT - match messages with the \\RECENT flag set
	SEEN - match messages that have been read (the \\SEEN flag is set)
	SINCE "date" - match messages with Date: after "date"
	SUBJECT "string" - match messages with "string" in the Subject:
	TEXT "string" - match messages with text "string"
	TO "string" - match messages with "string" in the To:
	UNANSWERED - match messages that have not been answered
	UNDELETED - match messages that are not deleted
	UNFLAGGED - match messages that are not flagged
	UNKEYWORD "string" - match messages that do not have the keyword "string"
	UNSEEN - match messages which have not been read yet

### Actions

Some other actions you would probably like to perform is moving an email and removing an email. The following figures shows basically the rest of the possible actions you can perform with Eden.

**Figure 8. Move Email to Another Folder**

	$imap->move(22363, 'Notes'); 

**Figure 9. Delete Email**

	$imap->remove(22363, true); 

**Figure 10. Don't Forget to Disconnect!**

	$imap->disconnect(); 


<a name="pop3"></a>
## POP3

*Eden's* POP3 object does uses `fsocket()` rather than PHP's built in IMAP functions for POP3. We chose to use `fsocket()` instead purely because of performance reasons. You'll find setting up POP3 is overall easier with *Eden*. `Figure 11` shows how to set up the POP3 object.

**Figure 11. Setting up the POP3 Object**

	$pop3 = eden('mail')->pop3(
		'pop.gmail.com', 
		'your_email@gmail.com', 
		'[YOUR PASSWORD]', 
		995, 
		true);

Very simply, there are four requirements that's needed and all requirements are dependent on your specific email provider. In the figure above we use GMAIL settings as an example. The last argument is a flag. Set it to true if your email provider requires SSL.Now that we have set the connection information, we can now continue to get a list of emails. `Figure 12` shows how we would go about doing that.

**Figure 12. Get Emails**

	$emails = $pop3->getEmails(0, 10); 
	$count = $pop3->getEmailTotal();

Executing the above figure would return you a list of emails as well as the total count. Some other 	actions you would probably like to perform is removing an email. The following figures shows basically the rest of the possible actions you can perform with *Eden*.

**Figure 13. Delete Email**

	$pop3->remove(100); 

**Figure 14. Don't Forget to Disconnect!**

	$pop3->disconnect(); 


<a name="smtp"></a>
## SMTP

SMTP in Eden is relatively easier. We use `fsocket()` rather than PHP's built in SMTP functions. We chose to use `fsocket()` because PHP's `mail()` function will most likely put your mail in the junk folder. SMTP works better because emails sent is sent by your actual email server versus your web host. `Figure 15` shows how to set up the SMTP object.

**Figure 15. Setting up the SMTP Object**

	$smtp = eden('mail')->smtp(
		'smtp.gmail.com', 
		'your_email@gmail.com', 
		'[YOUR PASSWORD]', 
		465, 
		true);

Very simply, there are four requirements that's needed and all requirements are dependent on your specific email provider. In the figure above we use GMAIL settings as an example. The last argument is a flag. Set it to true if your email provider requires SSL. The next part is simply send your email.

**Figure 16. Simply Send**

	$smtp->setSubject('Welcome!')
		->setBody('<p>Hello you!</p>', true)
		->setBody('Hello you!')
		->addTo('email1@gmail.com')
		->addTo('email2@gmail.com')
		->addCC('email3@gmail.com')
		->addCC('email4@gmail.com')
		->addBCC('email5@gmail.com')
		->addBCC('email6@gmail.com')
		->addAttachment('file.jpg', '/path/to/file.jpg', 'mime-type')
		->send();

In `Figure 16`, we basically laid out all the possible combinations of methods you can use to send email. It's important to set the `addTo()` method at least once.

**Figure 17. Don't Forget to Disconnect!**

	$smtp->disconnect(); 


====

#Contributing to Eden

##Setting up your machine with the Eden repository and your fork

1. Fork the main Eden repository (https://github.com/Eden-PHP/Mail)
2. Fire up your local terminal and clone the *MAIN EDEN REPOSITORY* (git clone git://github.com/Eden-PHP/Mail.git)
3. Add your *FORKED EDEN REPOSITORY* as a remote (git remote add fork git@github.com:*github_username*/Mail.git)

##Making pull requests

1. Before anything, make sure to update the *MAIN EDEN REPOSITORY*. (git checkout master; git pull origin master)
2. If PHP Unit testing is included in this package please make sure to update it and run the test to ensure everything still works (phpunit)
3. Once updated with the latest code, create a new branch with a branch name describing what your changes are (git checkout -b bugfix/fix-twitter-auth)
    Possible types:
    - bugfix
    - feature
    - improvement
4. Make your code changes. Always make sure to sign-off (-s) on all commits made (git commit -s -m "Commit message")
5. Once you've committed all the code to this branch, push the branch to your *FORKED EDEN REPOSITORY* (git push fork bugfix/fix-twitter-auth)
6. Go back to your *FORKED EDEN REPOSITORY* on GitHub and submit a pull request.
7. An Eden developer will review your code and merge it in when it has been classified as suitable.

CHANGELOG
* 3.0.0 (2012-01-30)

 * Version 3 release

Pop3.php
<?php //-->
/*
 * This file is part of the Mail package of the Eden PHP Library.
 * (c) 2013-2014 Openovate Labs
 *
 * Copyright and license information can be found at LICENSE
 * distributed with this package.
 */

namespace Eden\Mail;

/**
 * General available methods for common POP3 functionality
 *
 * @vendor Eden
 * @package Mail
 * @author Christian Blanquera cblanquera@openovate.com
 */
class Pop3 extends Base 
{
	const TIMEOUT = 30;
	const NO_SUBJECT = '(no subject)';
	
	protected $host	= NULL;
	protected $port	= NULL;
	protected $ssl = false;
	protected $tls = false;
	
	protected $username	= NULL;
	protected $password	= NULL;
	protected $timestamp = NULL;
	
	protected $socket = NULL;
	protected $loggedin	= false;
	private $debugging = false;
	
	/**
	 * Constructor - Store connection information
	 *
	 * @param string
	 * @param string
	 * @param string
	 * @param int|null
	 * @param bool
	 * @param bool
	 */
	public function __construct(
		$host, 
		$user, 
		$pass, 
		$port = NULL, 
		$ssl = false, 
		$tls = false
	) {
		Argument::i()
			->test(1, 'string')
			->test(2, 'string')
			->test(3, 'string')
			->test(4, 'int', 'null')
			->test(5, 'bool')
			->test(6, 'bool');
			
		if (is_null($port)) {
            $port = $ssl ? 995 : 110;
        }

		$this->host = $host;
		$this->username = $user;
		$this->password = $pass;
		$this->port = $port;
		$this->ssl = $ssl;
		$this->tls = $tls;
		
		$this->connect();
	}
	
	/**
	 * Connects to the server
	 *
	 * @return Eden\Mail\Pop3
	 */
	public function connect($test = false)
	{
		Argument::i()->test(1, 'bool');
		
		if($this->loggedin) {
			return $this;
		}
		
		$host = $this->host;
		
		if ($this->ssl) {
            $host = 'ssl://' . $host;
        }
		
		$errno  =  0;
        $errstr = '';

		$this->socket = fsockopen($host, $this->port, $errno, $errstr, self::TIMEOUT);
        
		if (!$this->socket) {
			//throw exception
			Exception::i()
				->setMessage(Exception::SERVER_ERROR)
				->addVariable($host.':'.$this->port)
				->trigger();
        }
		
		$welcome = $this->receive();

        strtok($welcome, '<');
        $this->timestamp = strtok('>');
        if (!strpos($this->timestamp, '@')) {
            $this->timestamp = null;
        } else {
            $this->timestamp = '<' . $this->timestamp . '>';
        }

        if ($this->tls) {
            $this->call('STLS');
            if (!stream_socket_enable_crypto(
				$this->socket, 
				true, 
				STREAM_CRYPTO_METHOD_TLS_CLIENT
			)) {
				$this->disconnect();
            	//throw exception
				Exception::i()
					->setMessage(Exception::TLS_ERROR)
					->addVariable($host.':'.$this->port)
					->trigger();
            }
        }
		
		if($test) {
			$this->disconnect();
			return $this;
		}
		
		//login
		if ($this->timestamp) {
            try {
                $this->call('APOP '.$this->username 
					. ' ' . md5($this->timestamp 
					. $this->password));
                return;
            } catch (Argument $e) {
                // ignore
            }
        }

        $this->call('USER '.$this->username);
        $this->call('PASS '.$this->password);
		
		$this->loggedin = true;
		
		return $this;
	}
	
	/**
	 * Disconnects from the server
	 *
	 * @return Eden\Mail\Pop3
	 */
	public function disconnect()
	{
		if (!$this->socket) {
            return $this;
        }

        try {
            $this->request('QUIT');
        } catch (Argument $e) {
            // ignore error - we're closing the socket anyway
        }

        fclose($this->socket);
        $this->socket = NULL;
		
		return $this;
	}
	
	/**
	 * Returns a list of emails given the range
	 *
	 * @param number start
	 * @param number range
	 * @return array
	 */
	public function getEmails($start = 0, $range = 10)
	{
		Argument::i()
			->test(1, 'int')
			->test(2, 'int');
		
		$total = $this->getEmailTotal();
		$total = $total['messages'];
		
		if($total == 0) {
			return array();
		}
		
        if (!is_array($start)) {
			$range = $range > 0 ? $range : 1;
			$start = $start >= 0 ? $start : 0;
			$max = $total - $start;

			if($max < 1) {
				$max = $total;
			}
			
			$min = $max - $range + 1;
			
			if($min < 1) {
				$min = 1;
			}
			
			$set = $min . ':' . $max; 
			
			if($min == $max) {
				$set = $min; 
			}
        }
		
		$emails = array();
		for($i = $min; $i <= $max; $i++) {
			$emails[] = $this->getEmailFormat($this->call('RETR '.$i, true));
		}
		
		return $emails;
    }
	
	/**
	 * Returns the total number of emails in a mailbox
	 *
	 * @return number
	 */
	public function getEmailTotal()
	{
		list($messages, $octets) = explode(' ', $this->call('STAT'));

		return $messages;	
	}
	
	/**
	 * Remove an email from a mailbox
	 *
	 * @param number uid
	 * @param string mailbox
	 * @return Eden\Mail\Pop3
	 */
	public function remove($msgno)
	{
		Argument::i()->test(1, 'int', 'string');
		
		$this->call("DELE $msgno");
		
		if(!$this->loggedin || !$this->socket) {
			return false;
		}
		
		if(!is_array($msgno)) {
			$msgno = array($msgno);
		}
		
		foreach($msgno as $number) {
			$this->call('DELE '.$number);
		
		}
		
		return $this;
	}
	
	/**
	 * Send it out and return the response
	 *
	 * @param string
	 * @param bool
	 * @return string|false
	 */
	protected function call($command, $multiline = false)
	{
		if(!$this->send($command)) {
			return false;
		}
		
		return $this->receive($multiline);
	}
	
	/**
	 * Returns the response when all of it is received
	 *
	 * @param string
	 * @return string
	 */
	protected function receive($multiline = false)
	{
		$result = @fgets($this->socket);
        $status = $result = trim($result);
        $message = '';
		
		if (strpos($result, ' ')) {
            list($status, $message) = explode(' ', $result, 2);
        }

        if ($status != '+OK') {
            return false;
        }

        if ($multiline) {
            $message = '';
            $line = fgets($this->socket);
            while ($line && rtrim($line, "\r\n") != '.') {
                if ($line[0] == '.') {
                    $line = substr($line, 1);
                }
				$this->debug('Receiving: '.$line);
                $message .= $line;
                $line = fgets($this->socket);
            };
        }

        return $message;
    }
	
	/**
	 * Sends out the command
	 * 
	 * @param string
	 * @return bool
	 */
	protected function send($command)
	{
		$this->debug('Sending: '.$command);
		
        return fputs($this->socket, $command . "\r\n");
   	}
	
	/**
	 * Debugging  
	 *
	 * @param string
	 * @return Eden\Mail\Imap
	 */
	private function debug($string)
	{
		if($this->debugging) {
			$string = htmlspecialchars($string);
			
			
			echo '<pre>'.$string.'</pre>'."\n";
		}
		return $this;
	}
	
	/**
	 * Secret Sauce - Transform an email string
	 * response to array key value format
	 *
	 * @param string
	 * @param array
	 * @return array
	 */
	private function getEmailFormat($email, array $flags = array())
	{
		//if email is an array
		if(is_array($email)) {
			//make it into a string
			$email = implode("\n", $email);
		}
		
		//split the head and the body
		$parts = preg_split("/\n\s*\n/", $email, 2);
		
		$head = $parts[0]; 
		$body = NULL;
		if(isset($parts[1]) && trim($parts[1]) != ')') {
			$body = $parts[1];
		}
		
		$lines = explode("\n", $head);
		$head = array();
		foreach($lines as $line) {
			if(trim($line) && preg_match("/^\s+/", $line)) {
				$head[count($head)-1] .= ' '.trim($line);
				continue;
			}
			
			$head[] = trim($line);
		}
		
		$head = implode("\n", $head);
		
		$recipientsTo = $recipientsCc = $recipientsBcc = $sender = array();
		
		//get the headers
		$headers1 	= imap_rfc822_parse_headers($head);
		$headers2 	= $this->getHeaders($head);
		
		//set the from
		$sender['name'] = NULL;
		if(isset($headers1->from[0]->personal)) {
			$sender['name'] = $headers1->from[0]->personal;
			//if the name is iso or utf encoded
			if(preg_match("/^\=\?[a-zA-Z]+\-[0-9]+.*\?/", strtolower($sender['name']))) {
				//decode the subject
				$sender['name'] = str_replace('_', ' ', mb_decode_mimeheader($sender['name']));
			}
		}
		
		$sender['email'] = $headers1->from[0]->mailbox . '@' . $headers1->from[0]->host;
		
		//set the to
		if(isset($headers1->to)) {
			foreach($headers1->to as $to) {
				if(!isset($to->mailbox, $to->host)) {
					continue;
				}
				
				$recipient = array('name'=>NULL);
				if(isset($to->personal)) {
					$recipient['name'] = $to->personal;
					//if the name is iso or utf encoded
					if(preg_match("/^\=\?[a-zA-Z]+\-[0-9]+.*\?/", strtolower($recipient['name']))) {
						//decode the subject
						$recipient['name'] = str_replace('_', ' ', mb_decode_mimeheader($recipient['name']));
					}
				}
				
				$recipient['email'] = $to->mailbox . '@' . $to->host;
				
				$recipientsTo[] = $recipient;
			}
		}
		
		//set the cc
		if(isset($headers1->cc)) {
			foreach($headers1->cc as $cc) {
				$recipient = array('name'=>NULL);
				if(isset($cc->personal)) {
					$recipient['name'] = $cc->personal;
					
					//if the name is iso or utf encoded
					if(preg_match("/^\=\?[a-zA-Z]+\-[0-9]+.*\?/", strtolower($recipient['name']))) {
						//decode the subject
						$recipient['name'] = str_replace('_', ' ', mb_decode_mimeheader($recipient['name']));
					}
				}
				
				$recipient['email'] = $cc->mailbox . '@' . $cc->host;
				
				$recipientsCc[] = $recipient;
			}
		}
		
		//set the bcc
		if(isset($headers1->bcc)) {
			foreach($headers1->bcc as $bcc) {
				$recipient = array('name'=>NULL);
				if(isset($bcc->personal)) {
					$recipient['name'] = $bcc->personal;				
					//if the name is iso or utf encoded
					if(preg_match("/^\=\?[a-zA-Z]+\-[0-9]+.*\?/", strtolower($recipient['name']))) {
						//decode the subject
						$recipient['name'] = str_replace('_', ' ', mb_decode_mimeheader($recipient['name']));
					}
				}
				
				$recipient['email'] = $bcc->mailbox . '@' . $bcc->host;
				
				$recipientsBcc[] = $recipient;
			}
		}
		
		//if subject is not set
		if(!isset( $headers1->subject ) || strlen(trim($headers1->subject)) === 0) {
			//set subject
			$headers1->subject = self::NO_SUBJECT;
		}
		
		//trim the subject
		$headers1->subject = str_replace(array('<', '>'), '', trim($headers1->subject));
		
		//if the subject is iso or utf encoded
		if(preg_match("/^\=\?[a-zA-Z]+\-[0-9]+.*\?/", strtolower($headers1->subject))) {
			//decode the subject
			$headers1->subject = str_replace('_', ' ', mb_decode_mimeheader($headers1->subject));
		}
		
		//set thread details
		$topic 	= isset($headers2['thread-topic']) ? $headers2['thread-topic'] : $headers1->subject;
		$parent = isset($headers2['in-reply-to']) ? str_replace('"', '', $headers2['in-reply-to']) : NULL;
		
		//set date
		$date = isset($headers1->date) ? strtotime($headers1->date) : NULL;
		
		//set message id
		if(isset($headers2['message-id'])) {
			$messageId = str_replace('"', '', $headers2['message-id']);
		} else {
			$messageId = '<eden-no-id-'.md5(uniqid()).'>';
		}
		
		$attachment = isset($headers2['content-type']) 
			&& strpos($headers2['content-type'], 'multipart/mixed') === 0;
		
		$format = array(
			'id'			=> $messageId,
			'parent'		=> $parent,
			'topic'			=> $topic,
			'mailbox'		=> 'INBOX',
			'date'			=> $date,
			'subject'		=> str_replace('’', '\'', $headers1->subject),
			'from'			=> $sender,
			'flags'			=> $flags,
			'to'			=> $recipientsTo,
			'cc'			=> $recipientsCc,
			'bcc'			=> $recipientsBcc,
			'attachment'	=> $attachment);
		
		if(trim($body) && $body != ')') {
			//get the body parts
			$parts = $this->getParts($email);
			
			//if there are no parts
			if(empty($parts)) {
				//just make the body as a single part
				$parts = array('text/plain' => $body);
			} 
			
			//set body to the body parts
			$body = $parts;
			
			//look for attachments
			$attachment = array();
			//if there is an attachment in the body
			if(isset($body['attachment'])) {
				//take it out
				$attachment = $body['attachment'];
				unset($body['attachment']);
			}
			
			$format['body']			= $body;
			$format['attachment']	= $attachment;
		}
		
		return $format;
	} 
	
	/**
	 * Returns email reponse headers
	 * array key value format
	 *
	 * @param string
	 * @return array
	 */
	private function getHeaders($rawData)
	{
		if(is_string($rawData)) {
			$rawData = explode("\n", $rawData);
		}
		
		$key = NULL;
		$headers = array();
		foreach($rawData as $line) {
			$line = trim($line);
			if(preg_match("/^([a-zA-Z0-9-]+):/i", $line, $matches)) {
				$key = strtolower($matches[1]);
				if(isset($headers[$key])) {
					if(!is_array($headers[$key])) {
						$headers[$key] = array($headers[$key]);
					}
					
					$headers[$key][] = trim(str_replace($matches[0], '', $line));
					continue;
				} 
				
				$headers[$key] = trim(str_replace($matches[0], '', $line));
				continue;
			} 
			
			if(!is_null($key) && isset($headers[$key])) {
				if(is_array($headers[$key])) {
					$headers[$key][count($headers[$key])-1] .= ' '.$line;	
					continue;
				}
				
				$headers[$key] .= ' '.$line; 
			}
		}
		
		return $headers;
	}
	
	/**
	 * Splits out body parts
	 * ie. plain, HTML, attachment
	 *
	 * @param string
	 * @param array
	 * @return array
	 */
	private function getParts($content, array $parts = array())
	{
		//separate the head and the body
		list($head, $body) = preg_split("/\n\s*\n/", $content, 2);
		//front()->output($head);
		//get the headers
		$head = $this->getHeaders($head);
		//if content type is not set
		if(!isset($head['content-type'])) {
			return $parts;
		}
		
		//split the content type
		if(is_array($head['content-type'])) {
			$type = array($head['content-type'][1]);
			if(strpos($type[0], ';') !== false) {
				$type = explode(';', $type[0], 2);
			}
		} else {
			$type = explode(';', $head['content-type'], 2);
		}
		
		//see if there are any extra stuff
		$extra = array();
		if(count($type) == 2) {
			$extra = explode('; ', str_replace(array('"', "'"), '', trim($type[1])));
		}
		
		//the content type is the first part of this
		$type = trim($type[0]);
		
		
		//foreach extra
		foreach($extra as $i => $attr) {
			//transform the extra array to a key value pair
			$attr = explode('=', $attr, 2);
			if(count($attr) > 1) {
				list($key, $value) = $attr;
				$extra[$key] = $value;
			}
			unset($extra[$i]);
		}
		
		//if a boundary is set
		if(isset($extra['boundary'])) {
			//split the body into sections
			$sections = explode('--'.str_replace(array('"', "'"), '', $extra['boundary']), $body);
			//we only want what's in the middle of these sections
			array_pop($sections);
			array_shift($sections);
			
			//foreach section
			foreach($sections as $section) {
				//get the parts of that
				$parts = $this->getParts($section, $parts);	
			}
		//if name is set, it's an attachment
		} else {
			
			//if encoding is set
			if(isset($head['content-transfer-encoding'])) {
				//the goal here is to make everytihg utf-8 standard
				switch(strtolower($head['content-transfer-encoding'])) {
					case 'binary':
						$body = imap_binary($body);
					case 'base64':
						$body = base64_decode($body);
						break;
					case 'quoted-printable':
						$body = quoted_printable_decode($body);
						break;
					case '7bit':
						$body = mb_convert_encoding ($body, 'UTF-8', 'ISO-2022-JP');
						break;
					default:
						$body = str_replace(array("\n", ' '), '', $body);
						break;
				}
			}
			
			if(isset($extra['name'])) {
				//add to parts
				$parts['attachment'][$extra['name']][$type] = $body;
			//it's just a regular body
			} else {
				//add to parts
				$parts[$type] = $body;
			}
		}
		return $parts;
	}
}